modin/pandas/dataframe.py

```diff
@@ -1161,15 +1168,11 @@ def insert(self, loc, column, value, allow_duplicates=False):  # noqa: PR01, D20
                 raise ValueError("Length of values does not match length of index")
             if not allow_duplicates and column in self.columns:
                 raise ValueError(f"cannot insert {column}, already exists")
-            if loc > len(self.columns):
+            if not -len(self.columns) <= loc <= len(self.columns):
```

Comment:
pandas.insert requires `0 <= loc <= len(df.columns)`. Why do we have support for negative indices here?

Reply:
A couple of lines below another exception is thrown - `raise ValueError("unbounded slice")` if loc < 0. That is, two different exceptions in the case when the `loc` is negative.

Reply:
I checked in pandas, and the case where loc is negative but greater than `-len(self.columns)` should result in the second error, so I think this code should just say this:
```suggestion
            if not loc <= len(self.columns):
```

Reply:
What version did you check on? On the latest release, the behavior is different.

```python
>>> import pandas as pd
>>> pd.__version__
'1.4.2'
>>> data = {"col1": [1, 2, 3], "col2": [2, 3, 4]}
>>> df = pd.DataFrame(data)
>>> df
   col1  col2
0     1     2
1     2     3
2     3     4
>>> df.insert(-1, "col3", df["col2"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\Miniconda3\envs\modin\lib\site-packages\pandas\core\frame.py", line 4445, in insert
    self._mgr.insert(loc, column, value)
  File "...\Miniconda3\envs\modin\lib\site-packages\pandas\core\internals\managers.py", line 1241, in insert
    bp = BlockPlacement(slice(loc, loc + 1))
  File "pandas\_libs\internals.pyx", line 56, in pandas._libs.internals.BlockPlacement.__cinit__
  File "pandas\_libs\internals.pyx", line 318, in pandas._libs.internals.slice_canonize
ValueError: unbounded slice
>>> df.insert(-3, "col3", df["col2"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\Miniconda3\envs\modin\lib\site-packages\pandas\core\frame.py", line 4445, in insert
    self._mgr.insert(loc, column, value)
  File "...\Miniconda3\envs\modin\lib\site-packages\pandas\core\internals\managers.py", line 1230, in insert
    new_axis = self.items.insert(loc, item)
  File "...\Miniconda3\envs\modin\lib\site-packages\pandas\core\indexes\base.py", line 6602, in insert
    new_values = np.insert(arr, loc, casted)
  File "<__array_function__ internals>", line 180, in insert
  File "...\Miniconda3\envs\modin\lib\site-packages\numpy\lib\function_base.py", line 5280, in insert
    raise IndexError(f"index {obj} is out of bounds for axis {axis} "
IndexError: index -3 is out of bounds for axis 0 with size 2
```

Reply:
I tried the exact same code snippet on 1.4.2 and I got a different error:
```python
Python 3.8.12 (default, Oct 12 2021, 06:23:56)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pandas as pd
pd.__version__

In [2]: pd.__version__
Out[2]: '1.4.2'

In [3]: data = {"col1": [1, 2, 3], "col2": [2, 3, 4]}

In [4]: df = pd.DataFrame(data)

In [5]: df.insert(-1, "col3", df["col2"])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-ecea5cb625c7> in <module>
----> 1 df.insert(-1, "col3", df["col2"])

~/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py in insert(self, loc, column, value, allow_duplicates)
   4443
   4444         value = self._sanitize_column(value)
-> 4445         self._mgr.insert(loc, column, value)
   4446
   4447     def assign(self, **kwargs) -> DataFrame:

~/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/internals/managers.py in insert(self, loc, item, value)
   1239             value = ensure_block_shape(value, ndim=self.ndim)
   1240
-> 1241         bp = BlockPlacement(slice(loc, loc + 1))
   1242         block = new_block_2d(values=value, placement=bp)
   1243

pandas/_libs/internals.pyx in pandas._libs.internals.BlockPlacement.__cinit__()

pandas/_libs/internals.pyx in pandas._libs.internals.slice_canonize()

ValueError: unbounded slice
```
<img width="1078" alt="image" src="https://user-images.githubusercontent.com/17012605/165394762-e3bd875a-1e8c-406d-b4b5-7b23a14eedd8.png">
maybe this is a python version thing? I'm on python 3.8.2

Reply:
@RehanSD with `loc=-1`, I also have this exception (`unbounded slice`). Another exception (`IndexError: index -3 is out of bounds for axis 0 with size 2`) I get when `loc=-3`.

Reply:
> Right, `df.insert()` raises different exception types starting from 1.4.0, since that version [this line](https://github.com/pandas-dev/pandas/blob/4bfe3d07b4858144c219b9346329027024102ab6/pandas/core/internals/managers.py#L1229-L1230) (which is basicaly an `Index.insert`) started to fail with an IndexError if the location is out `abs(idx_len)` bounds.
> 
> BTW, this overall if-elsing that's based on pandas internals that changes from version to version seems very unstable (we would again need to change this if pandas change something else there), I don't have though a good solution for how to make our code stable and preserve type-exceptions compatibility with pandas at the same time. For now the approach is ok, but we should start thinking in the direction of solving this exceptions-type compatibility problem.

I think before finding a solution to this problem, we should explicitly compare messages in exceptions during testing.